searchcore/src/vespa/searchcore/proton/matching/query.cpp

```diff
@@ -40,7 +40,25 @@ using std::vector;
 namespace proton::matching {
 
 namespace {
-void AddLocationNode(const string &location_str, Node::UP &query_tree, Location &fef_location) {
+
+Node::UP
+inject(Node::UP query, Node::UP to_inject) {
+    if (dynamic_cast<search::query::And *>(query.get())) {
+        dynamic_cast<search::query::And &>(*query).append(std::move(to_inject));
+    } else  if (dynamic_cast<search::query::Rank *>(query.get()) || dynamic_cast<search::query::AndNot *>(query.get())) {
+        search::query::Intermediate & root = dynamic_cast<search::query::Intermediate &>(*query);
+        root.prepend(inject(root.stealFirst(), std::move(to_inject)));
+    } else {
+        auto new_root = std::make_unique<ProtonAnd>();
+        new_root->append(std::move(to_inject));
```

Comment:
no need to invert the order of the backwards compatible case, append (instead of prepend) also better matches the AND case above.

Reply:
Agree, fixed